src/java/org/apache/cassandra/db/memtable/SkipListMemtable.java

```diff
@@ -0,0 +1,391 @@
+            right = null;
+
+        try
+        {
+            if (left == null)
+                return right == null ? partitions : partitions.headMap(right, includeRight);
+            else
+                return right == null
+                       ? partitions.tailMap(left, includeLeft)
+                       : partitions.subMap(left, includeLeft, right, includeRight);
+        }
+        catch (IllegalArgumentException e)
+        {
+            logger.error("Invalid range requested {} - {}", left, right);
+            throw e;
+        }
+    }
+
+    public Partition getPartition(DecoratedKey key)
+    {
+        return partitions.get(key);
+    }
+
+    public UnfilteredRowIterator iterator(DecoratedKey key, Slices slices, ColumnFilter selectedColumns, boolean reversed, SSTableReadsListener listener)
+    {
+        Partition p = getPartition(key);
+        if (p == null)
+            return null;
+        else
+            return p.unfilteredIterator(selectedColumns, slices, reversed);
+    }
+
+    public UnfilteredRowIterator iterator(DecoratedKey key)
+    {
+        Partition p = getPartition(key);
+        return p != null ? p.unfilteredIterator() : null;
+    }
+
+    private static int estimateRowOverhead(final int count)
+    {
+        // calculate row overhead
+        try (final OpOrder.Group group = new OpOrder().start())
+        {
+            int rowOverhead;
+            MemtableAllocator allocator = MEMORY_POOL.newAllocator("");
+            ConcurrentNavigableMap<PartitionPosition, Object> partitions = new ConcurrentSkipListMap<>();
+            final Object val = new Object();
+            for (int i = 0 ; i < count ; i++)
+                partitions.put(allocator.clone(new BufferDecoratedKey(new LongToken(i), ByteBufferUtil.EMPTY_BYTE_BUFFER), group), val);
+            double avgSize = ObjectSizes.measureDeep(partitions) / (double) count;
+            rowOverhead = (int) ((avgSize - Math.floor(avgSize)) < 0.05 ? Math.floor(avgSize) : Math.ceil(avgSize));
+            rowOverhead -= ObjectSizes.measureDeep(new LongToken(0));
+            rowOverhead += AtomicBTreePartition.EMPTY_SIZE;
+            rowOverhead += AbstractBTreePartition.HOLDER_UNSHARED_HEAP_SIZE;
+            allocator.setDiscarding();
+            allocator.setDiscarded();
+            return rowOverhead;
+        }
+    }
+
+    public FlushCollection<?> getFlushSet(PartitionPosition from, PartitionPosition to)
+    {
+        Map<PartitionPosition, AtomicBTreePartition> toFlush = getPartitionsSubMap(from, true, to, false);
+        long keySize = 0;
+
+        boolean trackContention = logger.isTraceEnabled();
+        if (trackContention)
+        {
+            int heavilyContendedRowCount = 0;
+
+            for (AtomicBTreePartition partition : toFlush.values())
+            {
+                keySize += partition.partitionKey().getKey().remaining();
+                if (trackContention && partition.useLock())
```

Comment:
Looking a bit more closely at this, it seems like the basic idea of `getFlushSet()` was separating `writeSortedContents()` into general logic and the skiplist-specific contention tracking stuff. That makes sense, but my only worry is that it looks like it also entails iterating over all the partitions twice. Would we be able to avoid that if the `FlushCollection` contract included something like a `beforeAppend()` callback that could do the contention tracking and logging?

Reply:
> I don't understand why trackContention would always be true

I think what he means is that `trackContention` isn't always true, but if it's true on line 264 (in the outer `if` statement) then it's also true on line 272 (in the inner `if` statement).

Reply:
> we previously iterated twice

My bad. I didn't see the `keySize` bit in the `FlushRunnable` constructor.

Reply:
Oh, I managed to miss that it's in that particular `if` too... sorry, fixed now.